### PR TITLE
Fix typing in utils.perf

### PR DESCRIPTION
**Pull request by dstansby**
# References and relevant issues

# Description
Fixes and adds typing to `napari.utils.perf.*`.

### CODE REVIEW

File: napari/utils/perf/_config.py
Comment (Czaki) [inline napari/utils/perf/_config.py:23]: It looks like this constructor call is obsolete. 
Also, there is a bad pattern because there is no `super().__init__(message)` call.

I thnk that we should use 

```python
class PerfmonConfigError(Exception):
    """Error parsing or interpreting config file."""

    @property 
    def message(self) -> str
        return self.args[0]
```

And maybe find all the places where `message` was used and use `args[0]` there. 
We may think about the usage of `ValueError` or `RuntimeError` as the base class.

File: napari/utils/perf/_config.py
Comment (Czaki) [inline napari/utils/perf/_config.py]:
```suggestion
from typing import Any, Callable, Dict, List, Optional, Tuple, Union
```

File: napari/utils/perf/_config.py
Comment (Czaki) [inline napari/utils/perf/_config.py]:
```suggestion
from types import ModuleType

def _patch_perf_timer(parent: Union[ModuleType, type], callable_name: str, label: str) -> None:
```
Of course, import should be moved on top.

File: napari/utils/perf/_config.py
Comment (Czaki) [inline napari/utils/perf/_config.py:44]: Why `instance` is used in the signature, but not in the body? It needs to be documented at least.

File: napari/utils/perf/_patcher.py
Comment (Czaki) [inline napari/utils/perf/_patcher.py:25]: Same as above

File: napari/utils/perf/_timers.py
Comment (Czaki) [inline napari/utils/perf/_timers.py:228]: Did we know anything about arguments and return type of callables?


Fast check suggest that it should something like:

```
Callable[[str, Optional[str], bool, ...], contextlib.AbstractContextManager],
```
and then reflect signatures of `add_instant_event` and `add_counter_event` 

With such improved type annotation, mypy could check if proper arguments are passed to returned functions.

**Review by Czaki — COMMENTED** [review]
I may need to do additional round of review, but the scope of adding typing to code is to create helpers that prevent us from making bugs on type checking phase, so we should avoid using `Any` or plain `Callable` if we know something about types. 

Sometimes it could also be better to refactor code instead of trying to type existing one.

**Review by Czaki — CHANGES_REQUESTED** [review]
I think thak this types needs improvements.

File: napari/utils/perf/_timers.py
Comment (dstansby) [reply napari/utils/perf/_timers.py:228]: These are tricky, if not impossible to type because of the `**kwargs` in all the funciton signatures. See https://docs.python.org/3/library/typing.html#annotating-callables:

>Callable cannot express complex signatures such as functions that take a variadic number of arguments, [overloaded functions](https://docs.python.org/3/library/typing.html#overload), or functions that have keyword-only parameters.

I also tried using `Protocol` classes to do this, but couldn't work out how to without mypy erroring.

I think having `Callable` here is still an improvement over no typing though?

File: napari/utils/perf/_config.py
Comment (dstansby) [reply napari/utils/perf/_config.py:44]: 🤷 , but since it's unused I think reasonable to type as `Any`?

File: napari/utils/perf/_config.py
Comment (dstansby) [reply napari/utils/perf/_config.py]: 👍 done

File: napari/utils/perf/_config.py
Comment (dstansby) [reply napari/utils/perf/_config.py:23]: I'd like to leave this PR as just typing the existing implementation instead of changing any of the code, if that's okay?

**Review by dstansby — COMMENTED** [review]
I've improved some of the `Any` typing, and left comments on the other inline comments.

File: napari/utils/perf/_config.py
Comment (Czaki) [reply napari/utils/perf/_config.py:23]: I prefer to improve the codebase instead of just plain typing. 

But let call some other core devs @napari/core-devs. If other has the same opinion as you then I will have no problem with it.

Comment (brisvag) [reply napari/utils/perf/_config.py:23]: I tend to agree; if adding types highlughts an issue, I would rather we leave this PR on hold, open another one where we fix the problem, and come back later!

If there's something more frustrating than fixing a weird old legacy piece of code, is having to do so with mypy screaming at you for things that you're trying to remove ^^'

Comment (dstansby) [reply napari/utils/perf/_config.py:23]: 👍 I'll convert this to draft then. Please ping me when it's ready to be reviewed again and I can rebase it.

**Review by Czaki — APPROVED** [review]

**Comment by Czaki** [comment]
I will merge this and then fix pointed problems in #6686

**Pull request merged**
